accepted/2023/feature-switch-attributes.md

```diff
@@ -0,0 +1,718 @@
+
+  On its own this is not enough for ILLink/ILCompiler to do branch elimination, because there's no tie to the feature switch name.
+
+- a reference to the feature name string:
+
+  ```csharp
+  class Feature {
+      [FeatureGuard("System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported")]
+      public static bool IsSupported => RuntimeFeature.IsDynamicCodeSupported;
+  }
+  ```
+
+  The analyzer would need to hard-code the fact that this string corresponds to `RequiresDynamicCodeAttribute`, unless we model this relationship via attributes that represent feature switches.
+
+  This would be sufficient for ILLink/ILCompiler to treat `IsSupported` as a constant based on the feature switch name.
+
+- a reference to the existing feature check property:
+
+  ```csharp
+  class Feature {
+      [FeatureGuard(typeof(RuntimeFeature), nameof(RuntimeFeature.IsDynamicCodeSupported))]
+      public static bool IsSupported => RuntimeFeature.IsDynamicCodeSupported;
+  }
+  ```
+
+  The analyzer would need to hard-code the fact that `RuntimeFeature.IsDynamicCodeSupported` corresponds to `RequiresDynamicCodeAttribute`, unless we model this relationship via attributes that represent feature switches.
+
+  This would be sufficient for ILLink/ILCompiler to treat `IsSupported` as a constant based on the feature switch name, assuming it has existing knowledge of the fact that `RuntimeFeature.IsDynamicCodeSupported` is controlled by the feature switch named `"System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported"`, either from a substitution XML or from a separate attribute that encodes this.
+
+### Danger of incorrect usage
+
+Since a feature guard silences warnings from the analyzer, there is some danger that `FeatureGuardAttribute` will be used carelessly as a way to silence warnings, even when the definition of the `IsSupported` property doesn't have any tie to the existing feature. For example:
+
+```csharp
+class Feature {
+    [FeatureGuard(typeof(RequiresDynamicCodeAttribute))]
+    public static bool SilenceWarnings => true; // BAD
+}
+```
+
+```csharp
+if (Feature.SilenceWarnings) {
+    APIWhichRequiresDynamicCode(); // No warnings
+}
+```
+
+This will silence warnings from the analyzer without any indication that `RequiresDynamicCode`-annotated code might be reached at runtime. We should be extremely clear in our documentation that this is not the intended use.
+
+### Validating correctness of feature guards
+
+Better would be for the analyzer to validate that the implementation of a feature check includes a check for `IsDynamicCodeSupported`. This could be done using the analyzer infrastructure we have in place without much cost for simple cases:
+
+```csharp
+class Feature {
+    [FeatureGuard(typeof(RequiresDynamicCodeAttribute))]
+    public static bool IsSupported => RuntimeFeature.IsDynamicCodeSupported && SomeOtherCondition(); // OK
```

Comment:
The above is perhaps okay, but the following wouldn't be.

```csharp
public static bool IsSupported => SomeOtherCondition() && RuntimeFeature.IsDynamicCodeSupported;
```

Consider the case where `SomeOtherCondition()` performs a check with side-effects. Do we have an opinion on that? Also what about with `||` instead of `&&`?

```csharp
public static bool IsSupported => RuntimeFeature.IsDynamicCodeSupported || SomeOtherCondition();
```

Reply:
Note that marking something as a FeatureGuard may not necessarily say anything about the trimmer behavior - I could see us implementing it such that the trimmer still has to call the property getter in that situation, unless it can prove that it's constant (through constant propagation).
This behavior would make a lot of sense even for NativeAOT - that doesn't do constant propagation, and wouldn't need to in the future. The scanner would be happy with just the branch removal, but leaving the getter call in place. And then it's up to JIT to inline/remove the getter call if it can prove it's constant.